# Домашнее задание: Управляющие конструкции

## Задание 1: Получение однозначного числа

Пользователь вводит целое число, программа складывает все цифры числа, с полученным числом — то же самое, и так до тех пор, пока не получится однозначное число.

<br>

**Примеры:**
* `545   -> 5`
* `12345 -> 6`

In [3]:
source_number = int(input("Введите целое число: "))

original_number = source_number

def sum_digits(n):
    digits = [int(d) for d in str(n)]
    return sum(digits)

while source_number > 9:
    source_number = sum_digits(source_number)

print(f"{original_number} -> {source_number}")


123 -> 6


## Задание 2: Кинотеатр

Дан список списков, каждый вложенный список состоит из 1 и 0, количество вложенных списков — количество рядов. Пользователь вводит, сколько билетов ему требуется. Программа должна найти ряд, где можно приобрести нужное количество билетов (места должны быть рядом). Если таких рядов несколько, то ближайший к экрану (ближайшим считается нулевой ряд). Если таких мест нет, то вывести False.

<br>

**Примеры:**
* `[[0,1,1,0], [1,0,0,0], [0,1,0,0]], 2 -> 1`
* `[[0,1,1,0], [1,0,1,0], [1,1,0,1]], 2 -> False`

In [7]:
seats_map = [[0,1,1,1,0], [1,0,0,0,1], [1,0,1,0,0]]

tickets = int(input("Сколько билетов рядом требуется? "))
found = False

for row_index, row in enumerate(seats_map):
    row_str = "".join(str(seats_map) for seats_map in row)

    if "0"*tickets in row_str:
        print (f"Есть {tickets} мест в ряду номер {row_index}")
        found = True
        break

if not found:
    print(False)


Есть 2 мест в ряду номер 1


## Задание 3: Алгоритм RLE

Необходимо написать упрощенную версию алгоритма RLE. 
Алгоритм RLE объединяет подряд идущие символы в коэффициент и символ.

<br>

**Примеры:**
* `asssdddsssddd -> 1a3s3d3s3d`
* `aaabbbbccccc  -> 3a4b5c`
* `abcba         -> 1a1b1c1b1a`

In [10]:
text = input("Введите строку: ")

rle_result = ""

if len(text) > 0:
    current_char = text[0]
    char_count = 1

    for char in text[1:]:
        if char == current_char:
            char_count += 1
        else:
            rle_result += str(char_count) + current_char
            current_char = char
            char_count = 1

    rle_result += str(char_count) + current_char

print(rle_result)

1a2s3d1f


## Задание 4: Шифр Цезаря

Реализуйте программу для шифрования текста с помощью шифра Цезаря.  
Шифр Цезаря — это метод шифрования, при котором каждая буква в тексте заменяется буквой, стоящей на фиксированное число позиций дальше в алфавите.

### Требования
1. Программа должна принимать на вход:
   - Строку текста для шифрования
   - Ключ шифра (число от 0 до 25)

2. Правила шифрования:
   - Шифруются только латинские буквы (a-z, A-Z)
   - Сохраняется регистр букв (заглавные остаются заглавными)
   - Пробелы, цифры и знаки препинания остаются без изменений
   - Сдвиг циклический (после 'z' идет 'a', после 'Z' идет 'A')

### Примеры

* `"Hello World", 3      -> "Khoor Zruog"`
* `"Python is great!", 5 -> "Udymts nx lwjfy!"`
* `"XYZ abc", 3          -> "ABC def"`
* `"Test 123", 1         -> "Uftu 123"`

### Алгоритм
* Для каждого символа в строке:
   - Если это латинская буква, найти её позицию в алфавите
   - Добавить к позиции значение ключа
   - Если результат больше 25, использовать остаток от деления на 26
   - Преобразовать обратно в букву, сохранив регистр
   - Если это не буква, оставить символ без изменений


In [11]:
text = input("Введите текст для шифрования: ")
key = int(input("Введите ключ шифра (0-25): "))

cipher_text = ""

for char in text:
    if char.isupper():
        pos = ord(char) - ord('A')
        new_pos = (pos + key) % 26
        cipher_text += chr(new_pos + ord('A'))

    elif char.islower():
        pos = ord(char) - ord('a')
        new_pos = (pos + key) % 26
        cipher_text += chr(new_pos + ord('a'))

    else:
        cipher_text += char

print(cipher_text)


BCD 100 bcd


## Задание 5: Табель успеваемости

Пользователь в бесконечном цикле (пока не будет введена пустая строка) вводит строки вида: 'название предмета' 'фамилия ученика' 'оценка'. После окончания ввода программа выводит в консоль название предмета, далее список учеников и все их оценки в виде таблицы.

<br>

**Примеры:**

<br>

*Ввод:*
```python
Математика Иванов  5
Математика Иванов  4
Литература Иванов  3
Математика Петров  5
Литература Сидоров 3
Литература Петров  5
Литература Иванов  4
Математика Сидоров 3
Математика Петров  5
```

<br>

*Результат:*
```python
Математика # вводим название предмета

# выводим список учеников и их оценки
Иванов  5 4
Петров  5 5
Сидоров 3
```

```python
Литература # вводим название предмета

# выводим список учеников и их оценки
Иванов  3 4
Сидоров 3
Петров  5
```

In [14]:
data = {}

while True:
    line = input("Введите предмет, фамилию и оценку (или пустая строка для конца): ")
    if line == "":
        break

    parts = line.split()
    if len(parts) != 3:
        print("Неправильный формат, попробуйте снова")
        continue

    subject, student, grade = parts[0], parts[1], parts[2]

    if subject not in data:
        data[subject] = {}

    if student not in data[subject]:
        data[subject][student] = []

    data[subject][student].append(grade)

for subject in data:
    print(subject)
    for student in data[subject]:
        grades = " ".join(data[subject][student])
        print(student, grades)


Математика
Иванов 4 3
Русский
Иванов 4
Петров 3
Сидоров 5
